File: services/controllers/interpolator.py
class Interpolator:

    def __init__(self):
        self.data = []

    def addIndexValue(self, index, value):
        self.data.append((index, value))

    def valueAtIndex(self, target_index):
        if target_index < self.data[0][0]:
            return None
        elif self.data[-1][0] < target_index:
            return None
        else:
            start = None
            end = None

            for (index, value) in self.data:
                if index == target_index:
                    return value
                else:
                    if index <= target_index:
                        start = (index, value)
                    elif target_index < index:
                        end = (index, value)
                        break

            index_delta = end[0] - start[0]
            percent = (target_index - start[0]) / index_delta
            value_delta = end[1] - start[1]

            return start[1] + value_delta * percent

    def to_array(self):
        result = []

        for (index, value) in self.data:
            result.append(index)
            result.append(value)

        return result

    def from_array(self, array):
        self.data = []
        for i in range(0, len(array), 2):
            self.addIndexValue(array[i], array[i + 1])


if __name__ == "__main__":
    pass

File: services/controllers/calibration/js/Interpolator.js
class Interpolator {
    constructor() {
        this.data = [];
    }

    addIndexValue(index, value) {
        this.data.push({index: index, value: value});

        // make sure items are in ascdending order by index
        this.data.sort((a, b) => a.index - b.index);
    }

    removeIndex(index) {
        for (var i = 0; i < this.data.length; i++) {
            if (this.data[i].index === index) {
                this.data.splice(i, 1);
                break;
            }
        }
    }

    valueAtIndex(target_index) {
        if (target_index < this.data[0].index || this.data[this.data.length - 1].index < target_index) {
            return null;
        }
        else {
            var start = null
            var end = null;

            for (var i = 0; i < this.data.length; i++) {
                let current = this.data[i];

                if (current.index === target_index) {
                    return current.value;
                }
                else {
                    if (current.index <= target_index) {
                        start = current;
                    }
                    else if (target_index < current.index) {
                        end = current;
                        break;
                    }
                }
            }

            let index_delta = end.index - start.index;
            let percent = (target_index - start.index) / index_delta;
            let value_delta = end.value - start.value;

            return start.value + value_delta * percent;
        }
    }

    to_array() {
        return this.data.reduce((accum, item) => {
            accum.push(item.index);
            accum.push(item.value);

            return accum;
        }, []);
    }
}
